lib/routes/account.js

```diff
@@ -1914,8 +1995,162 @@ module.exports = function (
+      config: {
+        auth: {
+          strategy: 'sessionTokenWithVerificationStatus'
+        },
+        validate: {
+          payload: {
+            email: validators.email().required()
+          }
+        },
+        response: {}
+      },
+      handler: function (request, reply) {
+        const sessionToken = request.auth.credentials
+        const uidHex = sessionToken.uid.toString('hex')
+        const userEmail = sessionToken.email
+        const ip = request.app.clientAddress
+        const email = request.payload.email
+        var emailCode
+
+        log.begin('Account.RecoveryEmailCreate', request)
+
+        customs.check(request, userEmail, 'createEmail')
+          .then(generateRandomValues)
+          .then(createEmail)
+          .then(sendEmailVerification)
+          .then(
+            function () {
+              reply({})
+            },
             reply
           )
+
+        function generateRandomValues() {
+          return random(16)
+            .then(bytes => {
+              emailCode = bytes
+            })
+        }
+
+        function createEmail() {
+          var emailData = {
+            email: email,
+            normalizedEmail: email.toLowerCase(),
+            emailCode: emailCode,
+            isVerified: false,
+            isPrimary: false,
+            uid: uidHex,
+            createdAt: Date.now()
+          }
+          return db.createEmail(uidHex, emailData)
+        }
+
+        function sendEmailVerification() {
+          return getGeoData(ip)
+            .then((geoData) => {
+              const account = {
+                uid: sessionToken.uid,
+                email: userEmail
+              }
+              return mailer.sendVerifySecondaryEmail(account, emailCode, userAgent.call({
+                acceptLanguage: request.app.acceptLanguage,
+                email: email,
+                userEmail: userEmail,
+                ip: ip,
+                location: geoData.location,
+                timeZone: geoData.timeZone
+              }, request.headers['user-agent'], log))
+            })
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/recovery_email/destroy',
+      config: {
+        auth: {
+          strategy: 'sessionTokenWithVerificationStatus'
+        },
+        validate: {
+          payload: {
+            email: validators.email().required()
+          }
+        },
+        response: {}
+      },
+      handler: function (request, reply) {
+        const sessionToken = request.auth.credentials
+        const uidHex = sessionToken.uid.toString('hex')
+        const userEmail = sessionToken.email
+        const email = request.payload.email
+
+        log.begin('Account.RecoveryEmailDestroy', request)
+
+        customs.check(request, userEmail, 'deleteEmail')
+          .then(deleteEmail)
+          .done(reply, reply)
+
+        function deleteEmail() {
+          return db.deleteEmail(uidHex, email.toLowerCase())
```

Comment:
As above, check consistency of hexlifying this here or internally in the db methods.